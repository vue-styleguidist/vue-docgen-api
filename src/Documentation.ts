import Map from 'ts-map'

export type BlockTag = ParamTag | Tag

export interface ParamType {
  name: string
  elements?: ParamType[]
}

export interface UnnamedParam {
  type?: ParamType
  description?: string | boolean
}

export interface Param extends UnnamedParam {
  name?: string
}

interface RootTag {
  title: string
}

export interface Tag extends RootTag {
  content: string | boolean
}

export interface ParamTag extends RootTag, Param {}

export interface DocBlockTags {
  description: string
  tags?: Array<ParamTag | Tag>
}

interface EventType {
  names: string[]
}

interface EventProperty {
  type: EventType
  name?: string
  description?: string | boolean
}

export interface EventDescriptor extends DocBlockTags {
  type?: EventType
  properties: EventProperty[] | undefined
}

export interface PropDescriptor {
  type?: { name: string; func?: boolean }
  description: string
  required?: string | boolean
  defaultValue?: { value: string; func?: boolean }
  tags: { [title: string]: BlockTag[] }
}

export interface MethodDescriptor {
  name: string
  description: string
  returns?: UnnamedParam
  tags?: { [key: string]: BlockTag[] }
  params?: Param[]
  modifiers: string[]
  [key: string]: any
}

export interface SlotDescriptor {
  description?: string
  bindings?: Record<string, any>
  scoped?: boolean
}

export interface ComponentDoc {
  displayName: string
  props: { [propName: string]: PropDescriptor } | undefined
  methods: MethodDescriptor[]
  slots: { [name: string]: SlotDescriptor }
  events?: { [name: string]: EventDescriptor }
  tags: { [key: string]: BlockTag[] }
  [key: string]: any
}

export class Documentation {
  private propsMap: Map<string, PropDescriptor>
  private methodsMap: Map<string, MethodDescriptor>
  private slotsMap: Map<string, SlotDescriptor>
  private eventsMap: Map<string, any>
  private dataMap: Map<string, any>
  constructor() {
    this.propsMap = new Map()
    this.methodsMap = new Map()
    this.slotsMap = new Map()
    this.eventsMap = new Map()

    this.dataMap = new Map()
  }

  public set(key: string, value: any) {
    this.dataMap.set(key, value)
  }

  public get(key: string): any {
    return this.dataMap.get(key)
  }

  public getPropDescriptor(propName: string): PropDescriptor {
    return this.getDescriptor(propName, this.propsMap, () => ({
      description: '',
      tags: {},
    }))
  }

  public getMethodDescriptor(methodName: string): MethodDescriptor {
    return this.getDescriptor(methodName, this.methodsMap, () => ({
      name: methodName,
      modifiers: [],
      description: '',
      tags: {},
    }))
  }

  public getEventDescriptor(eventName: string): EventDescriptor {
    return this.getDescriptor(eventName, this.eventsMap, () => ({
      properties: undefined,
      description: '',
      tags: [],
    }))
  }

  public getSlotDescriptor(slotName: string): SlotDescriptor {
    return this.getDescriptor(slotName, this.slotsMap, () => ({
      description: '',
    }))
  }

  public toObject(): ComponentDoc {
    const props = this.getPropsObject()
    const methods = this.getMethodsObject()
    const events = this.getEventsObject() || {}
    const slots = this.getSlotsObject()

    const obj: { [key: string]: any } = {}
    this.dataMap.forEach((value, key) => {
      if (key) {
        obj[key] = value
      }
    })

    return {
      ...obj,
      // initialize non null params
      description: obj.description || '',
      tags: obj.tags || {},

      // set all the static properties
      displayName: obj.displayName,
      props,
      events,
      methods,
      slots,
    }
  }

  private getDescriptor<T>(name: string, map: Map<string, T>, init: () => T): T {
    let descriptor = map.get(name)
    if (!descriptor) {
      map.set(name, (descriptor = init()))
    }
    return descriptor
  }

  private getObjectFromDescriptor<T>(map: Map<string, T>): { [name: string]: T } | undefined {
    if (map.size > 0) {
      const obj: { [name: string]: T } = {}
      map.forEach((descriptor, name) => {
        if (name && descriptor) {
          obj[name] = descriptor
        }
      })
      return obj
    } else {
      return undefined
    }
  }

  private getPropsObject(): { [propName: string]: PropDescriptor } | undefined {
    return this.getObjectFromDescriptor(this.propsMap)
  }

  private getMethodsObject(): MethodDescriptor[] {
    const methods: MethodDescriptor[] = []
    this.methodsMap.forEach((descriptor, name) => {
      if (name && methods && descriptor) {
        methods.push(descriptor)
      }
    })
    return methods
  }

  private getEventsObject(): { [eventName: string]: EventDescriptor } | undefined {
    return this.getObjectFromDescriptor(this.eventsMap)
  }

  private getSlotsObject(): { [slotName: string]: SlotDescriptor } {
    return this.getObjectFromDescriptor(this.slotsMap) || {}
  }
}
