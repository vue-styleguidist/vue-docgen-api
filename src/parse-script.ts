import { ParserPlugin } from '@babel/parser'
import * as bt from '@babel/types'
import { NodePath } from 'ast-types'
import Map from 'ts-map'
import buildParser from './babel-parser'
import { Documentation } from './Documentation'
import cacher from './utils/cacher'
import resolveExportedComponent from './utils/resolveExportedComponent'

// tslint:disable-next-line:no-var-requires
import recast = require('recast')

const ERROR_MISSING_DEFINITION = 'No suitable component definition found'

interface ParseScriptOptions {
  lang: 'ts' | 'js'
  filePath: string
  nameFilter?: string[]
}

export default function parseScript(
  source: string,
  documentation: Documentation,
  handlers: Array<
    (doc: Documentation, componentDefinition: NodePath, ast: bt.File, filePath: string) => void
  >,
  options: ParseScriptOptions,
) {
  const plugins: ParserPlugin[] = options.lang === 'ts' ? ['typescript'] : ['flow']

  const ast = cacher(() => recast.parse(source, { parser: buildParser({ plugins }) }), source)
  if (!ast) {
    throw new Error(ERROR_MISSING_DEFINITION)
  }

  const componentDefinitions = resolveExportedComponent(ast)

  if (componentDefinitions.size === 0) {
    throw new Error(ERROR_MISSING_DEFINITION)
  }

  executeHandlers(handlers, componentDefinitions, documentation, ast, options)
}

function executeHandlers(
  localHandlers: Array<
    (doc: Documentation, componentDefinition: NodePath, ast: bt.File, filePath: string) => void
  >,
  componentDefinitions: Map<string, NodePath>,
  documentation: Documentation,
  ast: bt.File,
  opt: ParseScriptOptions,
) {
  return componentDefinitions.forEach((compDef, name) => {
    if (compDef && name && (!opt.nameFilter || opt.nameFilter.indexOf(name) > -1)) {
      localHandlers.forEach(handler => handler(documentation, compDef, ast, opt.filePath))
    }
  })
}
