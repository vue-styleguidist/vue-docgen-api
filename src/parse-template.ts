import * as pug from 'pug'
import { ASTElement, ASTNode, compile, SFCBlock } from 'vue-template-compiler'
import { Documentation } from './Documentation'
import cacher from './utils/cacher'

export interface TemplateParserOptions {
  functional: boolean
  rootLeadingComment: string
}

export type Handler = (
  documentation: Documentation,
  templateAst: ASTElement,
  options: TemplateParserOptions,
) => void

export default function parseTemplate(
  tpl: SFCBlock,
  documentation: Documentation,
  handlers: Handler[],
  filePath: string,
) {
  if (tpl && tpl.content) {
    const template =
      tpl.attrs && tpl.attrs.lang === 'pug'
        ? pug.render(tpl.content, { filename: filePath })
        : tpl.content
    const ast = cacher(() => compile(template, { comments: true }).ast, template)
    const rootLeadingCommentArray = /^<!--(.+)-->/.exec(template.trim())
    const rootLeadingComment =
      rootLeadingCommentArray && rootLeadingCommentArray.length > 1
        ? rootLeadingCommentArray[1].trim()
        : ''
    if (ast) {
      traverse(ast, documentation, handlers, {
        functional: !!tpl.attrs.functional,
        rootLeadingComment,
      })
    }
  }
}

export function traverse(
  templateAst: ASTElement,
  documentation: Documentation,
  handlers: Handler[],
  options: TemplateParserOptions,
) {
  if (templateAst.type === 1) {
    handlers.forEach(handler => {
      handler(documentation, templateAst, options)
    })
    if (templateAst.children) {
      for (const childNode of templateAst.children) {
        if (isASTElement(childNode)) {
          traverse(childNode, documentation, handlers, options)
        }
      }
    }
  }
}

function isASTElement(node: ASTNode): node is ASTElement {
  return !!node && (node as ASTElement).children !== undefined
}
