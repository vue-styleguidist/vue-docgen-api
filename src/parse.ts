import * as fs from 'fs'
import * as path from 'path'
import { parseComponent, SFCDescriptor } from 'vue-template-compiler'
import { Documentation } from './Documentation'
import parseScript from './parse-script'
import parseTemplate from './parse-template'
import handlers from './script-handlers'
import templateHandlers from './template-handlers'
import cacher from './utils/cacher'

const ERROR_EMPTY_DOCUMENT = 'The passed source is empty'

/**
 * parses the source and returns the doc
 * @param {string} source code whose documentation is parsed
 * @param {string} filePath path of the current file against whom to resolve the mixins
 * @returns {object} documentation object
 */
export function parseFile(filePath: string, documentation: Documentation) {
  const source = fs.readFileSync(filePath, {
    encoding: 'utf-8',
  })
  return parseSource(source, filePath, documentation)
}

/**
 * parses the source and returns the doc
 * @param {string} source code whose documentation is parsed
 * @param {string} filePath path of the current file against whom to resolve the mixins
 * @returns {object} documentation object
 */
export function parseSource(source: string, filePath: string, documentation: Documentation) {
  const singleFileComponent = /\.vue$/i.test(path.extname(filePath))
  let parts: SFCDescriptor | null = null

  if (source === '') {
    throw new Error(ERROR_EMPTY_DOCUMENT)
  }

  if (singleFileComponent) {
    parts = cacher(() => parseComponent(source), source)
  }

  const scriptSource = parts ? (parts.script ? parts.script.content : undefined) : source
  if (scriptSource) {
    const lang =
      (parts && parts.script && parts.script.attrs && parts.script.attrs.lang === 'ts') ||
      /\.tsx?$/i.test(path.extname(filePath))
        ? 'ts'
        : 'js'
    parseScript(scriptSource, documentation, handlers, { lang, filePath })
  }

  // get slots from template
  if (parts && parts.template) {
    parseTemplate(parts.template, documentation, templateHandlers, filePath)
  }

  if (!documentation.get('displayName')) {
    // a component should always have a display name
    documentation.set('displayName', path.basename(filePath).replace(/\.\w+$/, ''))
  }
}
