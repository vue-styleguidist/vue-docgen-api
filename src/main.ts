import { ComponentDoc, Documentation } from './Documentation'
import { parseFile, ParseOptions, parseSource as parseSourceLocal } from './parse'

export { ComponentDoc, ParseOptions }

export function parse(
  filePath: string,
  opts?: ParseOptions | { [alias: string]: string },
): ComponentDoc {
  const doc = new Documentation()
  const options: ParseOptions = isParserOptions(opts) ? opts : { filePath, aliases: opts }
  parseFile(doc, options)
  return doc.toObject()
}

export function parseSource(
  source: string,
  filePath: string,
  opts?: ParseOptions | { [alias: string]: string },
): ComponentDoc {
  const doc = new Documentation()
  const options: ParseOptions = isParserOptions(opts) ? opts : { filePath, aliases: opts }
  parseSourceLocal(doc, source, options)
  return doc.toObject()
}

function isParserOptions(opts: any): opts is ParseOptions {
  return !!opts && !!opts.aliases
}
