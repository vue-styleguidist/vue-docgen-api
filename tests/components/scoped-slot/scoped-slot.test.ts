import * as path from 'path'
import { ComponentDoc } from '../../../src/Documentation'
import { parse } from '../../../src/main'

const button = path.join(__dirname, './Wrapper.vue')
let docButton: ComponentDoc

describe('tests wrapper with root slot', () => {
  beforeEach(done => {
    docButton = parse(button)
    done()
  })

  it('should have a slot.', () => {
    expect(Object.keys(docButton.slots).length).toEqual(1)
  })

  it('should have a wrapper slot.', () => {
    expect(docButton.slots.footer.description).toBe('Modal footer here')
  })

  it('should show the slot as scoped', () => {
    expect(docButton.slots.footer.scoped).toBeTruthy()
  })

  it('should match the reference for the footer slot', () => {
    expect(docButton.slots.footer).toMatchSnapshot()
  })
})
