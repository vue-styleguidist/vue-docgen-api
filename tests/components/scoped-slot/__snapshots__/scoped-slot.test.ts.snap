// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests wrapper with root slot should match the reference for the footer slot 1`] = `
Object {
  "bindings": Object {
    ":item": "item",
  },
  "description": "Modal footer here",
  "scoped": true,
}
`;
